strawberryfields/utils.py

```diff
@@ -673,3 +691,334 @@ def f_proxy(*args):
+    ``D`` is the dimension of each vectorized index (i.e. for :math:`N` modes, ``D=cutoff_dim**N``):
+
+    >>> choi_matrix = liouville.reshape(D**2, D**2).T
+    >>> liouville_matrix = choi.reshape(D**2, D**2).T
+
+    **Kraus representation**
+
+    Adjacent indices of each Kraus operator correspond to output-input pairs of the same mode, so the action
+    of the channel can be written as (here for one mode or for ``vectorize_modes=True``):
+
+    >>> rho_out = np.einsum('abc,cd,aed->be', kraus, rho_in, np.conj(kraus))
+
+    Notice the transpose on the third index string (``aed`` rather than ``ade``), as the last operator should be the
+    conjugate transpose of the first, and we cannot just do ``np.conj(kraus).T`` because ``kraus`` has 3 indices and we
+    just need to transpose the last two.
+
+    Args:
+        engine (Engine): the engine containing the circuit
+        cutoff_dim (int): dimension of each index
+        representation (str): choice between ``'choi'``, ``'liouville'`` or ``'kraus'``
+        vectorize_modes (bool): if True, reshapes the result into rank-4 tensor,
+            otherwise it returns a rank-4N tensor, where N is the number of modes.
+
+    Returns:
+        The numerical array of the channel, according to the specified representation
+        and vectorization options.
+
+    Raises:
+        TypeError: if the gates used to construct the circuit are not all unitary or channels.
+    """
+    # if is_unitary(engine):
+    #     #raise Warning(f"This circuit is unitary and you could use extract_unitary for a more compact representation")
+    #     pass
+
+    if not is_channel(engine):
+        raise TypeError("The circuit definition contains elements that are neither of type Gate nor of type Channel")
+
+    from copy import deepcopy
+
+    _engine = _engine_with_CJ_cmd_queue(deepcopy(engine), cutoff_dim=cutoff_dim)
+    N = _engine.init_num_subsystems
+
+    backend = load_backend('fock')
+    backend.begin_circuit(num_subsystems=_engine.num_subsystems, cutoff_dim=cutoff_dim, hbar=_engine.hbar, pure=True)
+    choi = _engine.run(backend, cutoff_dim=cutoff_dim).dm()
+    choi = np.einsum('abcd->cdab', _vectorize_dm(choi))
+
+    if representation.lower() == 'choi':
+        result = choi
+        if not vectorize_modes:
+            result = _unvectorize_dm(result, N)
+
+    elif representation.lower() == 'liouville':
+        result = np.einsum('abcd -> dbca', choi)
+        if not vectorize_modes:
+            result = _unvectorize_dm(result, N)
+
+    elif representation.lower() == 'kraus':
+        eigvals, eigvecs = np.linalg.eig(np.einsum('abcd -> cadb', choi).reshape([cutoff_dim**(2*N), cutoff_dim**(2*N)]))
+        eigvecs = eigvecs[:, ~np.isclose(abs(eigvals), 0)]
+        eigvals = eigvals[~np.isclose(abs(eigvals), 0)]
+        result = np.einsum('abc->cab', np.einsum('b,ab->ab', np.sqrt(eigvals), eigvecs).reshape([cutoff_dim**N, cutoff_dim**N, -1]))
```

Comment:
@ziofil, your docstring is fantastic, but could you also add some comments here (especially to the Kraus lines) just to clarify what is happening on each line?

Reply:
Done